lib/service/Service.ts

```diff
@@ -330,6 +330,34 @@ class Service extends EventEmitter {
     return this.orderBook.removePair(pairId);
   }
 
+  public executeSwap = async (args: { orderId: string, pairId: string, peerPubKey: string, quantity: number }): Promise<SwapResult> => {
+    if (!this.orderBook.nomatching) {
+      throw errors.NOMATCHING_MODE_IS_REQUIRED();
+    }
+
+    const { orderId, pairId, peerPubKey } = args;
+    const quantity = args.quantity > 0 ? args.quantity : undefined; // passing 0 quantity will work fine, but it's prone to bugs
+
+    const maker = this.orderBook.removePeerOrder(orderId, pairId, peerPubKey, quantity).order;
+    const taker = this.orderBook.stampOwnOrder({
+      localId: '',
+      pairId,
+      price: maker.price,
+      isBuy: !maker.isBuy,
+      quantity: quantity || maker.quantity
+    });
+
+    try {
+      const swapResult = await this.swaps.executeSwap(maker, taker);
+      this.emit('peerOrder.filled', maker);
+      return swapResult;
+    } catch (err) {
+      this.emit('peerOrder.invalidation', maker);
+      // TODO: penalize peer for failed swap? penalty severity should depend on reason for failure
```

Comment:
Emitting these events and penalizing the peer from `Service` is a problem, because nothing is listening to `Service` for these events. This logic is copied from the `placeOrder` (around line 311) method in OrderBook.ts, so I figure we should move that into is own method and call that method directly from here. Maybe `OrderBook.executeSwap`?

Reply:
Yes, I was thinking of putting it inside `OrderBook`, to be used also from `placeOrder`, but the latter has some other logic as well and I thought it would be confusing to separate it from its flow. 

But I overlooked events issue you mentioned, so obviously it does need to be inside `OrderBook`.

I ended up putting it inside `OrderBook`, but not to be used from `placeOrder`, so code is duplicated. Let me know if it looks fine to you. Otherwise I can try to make it work without code duplication.

Reply:
Yeah I do think it would be better without the code duplication. I'll make a separate comment for what I have in mind.